Do not cache all token check failures

johnswanson opened the pull request:
We want to cache token checks to avoid an issue where we repeatedly ask the store "hey, is this token valid?? is this token valid?? is this token valid??" for the same token.

However, transient errors can also occur. For example, maybe a network issue causes the HTTP request to fail entirely. In this case, if we cache the result, the user needs to restart metabase (or wait 5 minutes until the cache is cleared) before they can attempt to validate their token again.

This PR moves the cache logic deeper into the stack. We want to cache "successful" responses from the store API - cases where the store has told us categorically that the token is or is not valid. We don't want or need to cache other things that might happen. Maybe your token isn't the right length - we can recalculate that, it's ok. Maybe you get a 503 error from the Store - we should let you retry. Maybe your network is having issues and you can't contact the Store at all - again, we should let you retry.

~The one potential issue I see here is that if the store goes down temporarily, we'll massively increase the number of requests we send to the store, potentially making it harder to recover from an outage. If this is a concern, I can add a circuit breaker: if we repeatedly get errors back from the store, back off and stop making requests for a while.~

Edit: I added the following:

In the pathological case where the store goes down for >5 minutes, the
cache will expire and all instances everywhere will start repeatedly
making requests for token validation at once. This might make recovering
from an outage more difficult.

This adds a circuit breaker around the API request. If the call
repeatedly throws (5XX errors, socket timeouts, etc.) then we'll pause
for 30 seconds, during which time all calls to token validation will
immediately fail without making any request to the API.

After 30 seconds, we'll allow one request through to the API. If it
succeeds, we'll go back to normal operation. Otherwise, we'll wait
another 30 seconds.


Fixes #41654

johnswanson committed:
Do not cache all token check failures

We want to cache token checks to avoid an issue where we repeatedly ask
the store "hey, is this token valid?? is this token valid?? is this
token valid??" for the same token.

However, transient errors can also occur. For example, maybe a network
issue causes the HTTP request to fail entirely. In this case, if we
cache the result, the user needs to restart metabase (or wait 5 minutes
until the cache is cleared) before they can attempt to validate their
token again.

This PR moves the cache logic deeper into the stack. We want to cache
"successful" responses from the store API - cases where the store has
told us categorically that the token is or is not valid. We don't want
or need to cache other things that might happen. Maybe your token isn't
the right length - we can recalculate that, it's ok. Maybe you get a 503
error from the Store - we should let you retry. Maybe your network is
having issues and you can't contact the Store at all - again, we should
let you retry.

The one potential issue I see here is that if the store goes down, we'll
massively increase the number of requests we send to the store,
potentially making it harder to recover. If this is a concern, I can add
a circuit breaker: if we repeatedly get errors back from the store, back
off and stop making requests for a while.

johnswanson commented on src/metabase/public_settings/premium_features.clj:
```diff
@@ -194,28 +214,11 @@
-  (let [lock (Object.)
-        f    (memoize/ttl
-              (fn [token]
-                ;; this is a sanity check to make sure we can actually get the active user count BEFORE we try to call
```
Note: I removed this because I moved caching deeper - it only wraps the HTTP request itself now. Therefore, a failure to get the active users count will no longer be cached.

noahmoss commented on src/metabase/public_settings/premium_features.clj:
```diff
@@ -110,18 +110,38 @@
 (defn- fetch-token-and-parse-body
   [token base-url site-uuid]
-  (let [fut    (future (fetch-token-and-parse-body* token base-url site-uuid))
+  (let [active-user-count (cached-active-users-count)
+        fut    (future (fetch-token-and-parse-body* token base-url site-uuid active-user-count))
```
```suggestion
  (let [active-user-count (cached-active-users-count)
        fut (future (fetch-token-and-parse-body* token base-url site-uuid active-user-count))
```

johnswanson committed:
Add a circuit breaker to store API requests

In the pathological case where the store goes down for >5 minutes, the
cache will expire and all instances everywhere will start repeatedly
making requests for token validation at once. This might make recovering
from an outage more difficult.

This adds a circuit breaker around the API request. If the call
repeatedly throws (5XX errors, socket timeouts, etc.) then we'll pause
for 1 minute, during which time all calls to token validation will
immediately fail without making any request to the API.

After one minute, we'll allow one request through to the API. If it
succeeds, we'll go back to normal operation. Otherwise, we'll wait
another minute.

noahmoss reviewed (approved):
Circuit breaker lgtm!
